File: IntegrationGuideForAndroid/Apps/trafficAuthentication/src/main/java/com/example/integrationguide/OkHttpStack.java
/*
 * Copyright (c) $2018 AirWatch, LLC. All rights reserved.
 * This product is protected by copyright and intellectual property laws in
 * the United States and other countries as well as by international treaties.
 * AirWatch products may be covered by one or more patents listed at
 * http://www.vmware.com/go/patents.
 */

package com.example.integrationguide;

import android.util.Log;

import com.airwatch.gateway.clients.AWOkHttpClient;
import com.android.volley.AuthFailureError;
import com.android.volley.Header;
import com.android.volley.Request;
import com.android.volley.toolbox.BaseHttpStack;
import com.android.volley.toolbox.HttpResponse;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import okhttp3.Headers;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class OkHttpStack extends BaseHttpStack {

    private String TAG = OkHttpStack.class.getName();
    private OkHttpClient mClient;

    public OkHttpStack(OkHttpClient client) {
        mClient = client;
    }

    @Override
    public HttpResponse executeRequest(Request<?> request, Map<String, String> additionalHeaders) throws IOException, AuthFailureError {

        //Set OkHttpClient new builder
        OkHttpClient.Builder clientBuilder = mClient.newBuilder();
        int timeoutMs = request.getTimeoutMs();

        //Set Volley request timer and set up the url
        clientBuilder.connectTimeout(timeoutMs, TimeUnit.MILLISECONDS);
        clientBuilder.readTimeout(timeoutMs, TimeUnit.MILLISECONDS);
        clientBuilder.writeTimeout(timeoutMs, TimeUnit.MILLISECONDS);

        okhttp3.Request.Builder okHttpRequestBuilder = new okhttp3.Request.Builder();
        okHttpRequestBuilder.url(request.getUrl());

        //Include headers in OkHttpRequest
        Map<String, String> headers = request.getHeaders();
        for (final String name : headers.keySet()) {
            okHttpRequestBuilder.addHeader(name, headers.get(name));
        }

        for (final String name : additionalHeaders.keySet()) {
            okHttpRequestBuilder.addHeader(name, additionalHeaders.get(name));
        }

        //Execute the okHttpRequest and get the response
        OkHttpClient client = clientBuilder.build();
        okhttp3.Request okHttpRequest = okHttpRequestBuilder.build();

        Response okHttpResponse = AWOkHttpClient.newCall(client,okHttpRequest).execute();

        Log.d(TAG,"okHttpResponse "+okHttpResponse.toString());

        //Convert the response to com.android.volley.toolbox.HttpResponse
        int code = okHttpResponse.code();
        ResponseBody body = okHttpResponse.body();
        InputStream content = body == null ? null : body.byteStream();
        int contentLength = body == null ? 0 : (int) body.contentLength();
        List<Header> responseHeaders = mapHeaders(okHttpResponse.headers());
        return new HttpResponse(code, responseHeaders, contentLength, content);
    }

    private List<Header> mapHeaders(Headers responseHeaders) {
        List<Header> headers = new ArrayList<>();
        for (int i = 0, len = responseHeaders.size(); i < len; i++) {
            final String name = responseHeaders.name(i), value = responseHeaders.value(i);
            if (name != null) {
                headers.add(new Header(name, value));
            }
        }
        return headers;
    }

}

File: IntegrationGuideForAndroid/Apps/tunnel/src/main/java/com/example/integrationguide/OkHttpStack.kt
// Copyright 2023 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

package com.example.integrationguide

import android.util.Log
import okhttp3.OkHttpClient
import com.android.volley.toolbox.BaseHttpStack
import kotlin.Throws
import com.android.volley.AuthFailureError
import com.airwatch.gateway.clients.AWOkHttpClient
import com.android.volley.Header
import com.android.volley.Request
import com.android.volley.toolbox.HttpResponse
import okhttp3.Headers
import java.io.IOException
import java.util.ArrayList
import java.util.concurrent.TimeUnit

class OkHttpStack(private val mClient: OkHttpClient) : BaseHttpStack() {

    companion object {
        private val TAG = OkHttpStack::class.java.name
    }

    @Throws(IOException::class, AuthFailureError::class)
    override fun executeRequest(
        request: Request<*>,
        additionalHeaders: Map<String, String>,
    ): HttpResponse {

        //Set OkHttpClient new builder
        val clientBuilder: OkHttpClient.Builder = mClient.newBuilder()
        val timeoutMs = request.timeoutMs

        //Set Volley request timer and set up the url
        clientBuilder.connectTimeout(timeoutMs.toLong(), TimeUnit.MILLISECONDS)
        clientBuilder.readTimeout(timeoutMs.toLong(), TimeUnit.MILLISECONDS)
        clientBuilder.writeTimeout(timeoutMs.toLong(), TimeUnit.MILLISECONDS)
        val okHttpRequestBuilder = okhttp3.Request.Builder()
        okHttpRequestBuilder.url(request.url)

        //Include headers in OkHttpRequest
        val headers = request.headers
        for (name in headers.keys) {
            okHttpRequestBuilder.addHeader(name, headers[name]!!)
        }
        for (name in additionalHeaders.keys) {
            okHttpRequestBuilder.addHeader(name, additionalHeaders[name]!!)
        }

        //Execute the okHttpRequest and get the response
        val client: OkHttpClient = clientBuilder.build()
        val okHttpRequest: okhttp3.Request = okHttpRequestBuilder.build()
        val okHttpResponse = AWOkHttpClient.newCall(client, okHttpRequest).execute()
        Log.d(TAG, "okHttpResponse $okHttpResponse")

        //Convert the response to com.android.volley.toolbox.HttpResponse
        val code = okHttpResponse.code
        val body = okHttpResponse.body
        val content = body?.byteStream()
        val contentLength = body?.contentLength()?.toInt() ?: 0
        val responseHeaders = mapHeaders(okHttpResponse.headers)
        return HttpResponse(code, responseHeaders, contentLength, content)
    }

    private fun mapHeaders(responseHeaders: Headers): List<Header> {
        val headers: MutableList<Header> = ArrayList()
        var i = 0
        val len = responseHeaders.size
        while (i < len) {
            val name = responseHeaders.name(i)
            val value = responseHeaders.value(i)
            headers.add(Header(name, value))
            i++
        }
        return headers
    }
}